### PR TITLE
S3authservice

### PR DESCRIPTION
**Pull request by mjarco**

### CODE REVIEW

File: glide.yaml [hunk 1]
```diff
@@ -21,6 +21,7 @@ import:
 - package: gopkg.in/tylerb/graceful.v1
   version: ^1.2.15
 - package: gopkg.in/yaml.v2
+- package: github.com/levigross/grequests
```
Comment (janisz) [inline glide.yaml]: Missing version.
Do you really need this dependency to perform one `Get`?

File: storages/auth/s3_v2.go [hunk 2]
```diff
@@ -250,15 +251,50 @@ type signRoundTripper struct {
...
 func SignDecorator(keys Keys) httphandler.Decorator {
 	return func(rt http.RoundTripper) http.RoundTripper {
 		return signRoundTripper{rt: rt, keys: keys}
 	}
 }
+
+// SignAuthServiceDecorator will compute
+func SignAuthServiceDecorator(backend, endpoint string) httphandler.Decorator {
+	return func(rt http.RoundTripper) http.RoundTripper {
+		return signAuthServiceRoundTripper{rt: rt, backend: backend, crd: crdstore.GetInstance(endpoint)}
```
Comment (janisz) [inline storages/auth/s3_v2.go]: `GetInstance` is not thread-safe

File: transport/crdstore/crdstore.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	if value, ok := cs.cache.Load(key); ok {
+		csd = value.(*CredentialsStoreData)
+	}
+
+	switch {
+	case csd == nil || csd.AccessKey == "":
+		return cs.updateCache(accessKey, backend, key, csd, true)
+	case time.Now().After(csd.EOL):
+		return cs.updateCache(accessKey, backend, key, csd, false)
+	case time.Now().Add(cs.TTL / time.Second * (100 - refreshTTLPercent) * 10 * time.Millisecond).After(csd.EOL):
```
Comment (janisz) [inline transport/crdstore/crdstore.go]: Magic numbers!

File: storages/auth/auth.go [hunk 4]
```diff
@@ -38,4 +40,17 @@ var Decorators = map[string]func(map[string]string) (httphandler.Decorator, erro
 		}
 		return SignDecorator(k), nil
 	},
+	S3AuthService: func(extra map[string]string) (httphandler.Decorator, error) {
```
Comment (janisz) [inline storages/auth/auth.go]: Can you rename `extra` with something meaningful?

File: storages/auth/s3_v2.go [hunk 5]
```diff
@@ -250,15 +251,50 @@ type signRoundTripper struct {
...
 
+// RoundTrip implements http.RoundTripper interface
+func (srt signAuthServiceRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	accessKey := strings.Trim(req.Header.Get("Authorization"), " ")
+	start := strings.IndexAny(accessKey, " ")
+	end := strings.IndexAny(accessKey, ":")
+	if start == -1 || end == -1 {
+		return &http.Response{StatusCode: http.StatusBadRequest, Request: req}, fmt.Errorf("cannot find AWS AccessKey in request")
+	}
+	accessKey = accessKey[start+1 : end]
```
Comment (janisz) [inline storages/auth/s3_v2.go]: Extract code above to separate pure function that will extract access key from a header.

Comment (bieli) [reply storages/auth/s3_v2.go]: you can use `strings.SplitN` for extracting values from header `Authorization` - example: [https://gist.github.com/sambengtson/bc9f76331065f09e953f]

File: storages/auth/s3_v2.go (hunk 2 above)
Comment (bnogas) [reply storages/auth/s3_v2.go]: it doesn't have to be

File: config/validator.go [hunk 6]
```diff
@@ -61,52 +61,53 @@ func UniqueValuesInSliceValidator(v interface{}, param string) error {
...
-	// 	errorsList := make(map[string][]error)
-	// 	errorsList["RegionsEntryLogicalValidator"] = errList
-	// 	*validationErrors = mergeErrors(*validationErrors, errorsList)
-	// } else {
-	// 	*valid = true
-	// }
+	errList := make([]error, 0)
+	if len(c.Regions) == 0 {
+		errList = append(errList, errors.New("Empty regions definition"))
+	} else {
```
Comment (janisz) [inline config/validator.go]: This `else` is redundant. If `c.Regions` is empty loop will not be executed.

File: config/validator.go [hunk 7]
```diff
@@ -61,52 +61,53 @@ func UniqueValuesInSliceValidator(v interface{}, param string) error {
...
+		for regionName, regionConf := range c.Regions {
+
+			if len(regionConf.Clusters) == 0 {
+				errList = append(errList, fmt.Errorf("No clusters defined for region \"%s\"", regionName))
+			}
+			for _, regionCluster := range regionConf.Clusters {
```
Comment (janisz) [inline config/validator.go]: Extract body of this loop to a well-named function e.g., `ClusterValidator()`

File: config/validator.go [hunk 8]
```diff
@@ -61,52 +61,53 @@ func UniqueValuesInSliceValidator(v interface{}, param string) error {
...
-	// }
-	// *validationErrors = mergeErrors(*validationErrors, errorsList)
+	if listenParts[0] == listenTechnicalParts[0] && listenParts[1] == listenTechnicalParts[1] {
+		*valid = false
+		errorDetail := []error{errors.New("Listen and TechnicalEndpointListen has the same port")}
+		errorsList["ListenPortsLogicalValidator"] = errorDetail
+	} else {
+		*valid = true
+	}
+	*validationErrors = mergeErrors(*validationErrors, errorsList)
```
Comment (janisz) [inline config/validator.go]: Move this to `if`. `errorsList` could be `if`-scoped

File: crdstore/config/config.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import (
+	"time"
+
+	"github.com/allegro/akubra/types"
+)
+
+// CredentialsStore - config for credential store
+type CredentialsStore struct {
+	Endpoint types.YAMLUrl `yaml:"Endpoint"`
+	TTL      time.Duration `yaml:"TTL"`
```
Comment (janisz) [inline crdstore/config/config.go]: Is `time.Duration` serialized in human readable string or nanoseconds?

Comment (bnogas) [reply crdstore/config/config.go]: I've changed it to metrics.Interval so it's human reable now

File: config/validator.go (hunk 7 above)
Comment (mjarco) [reply config/validator.go]: thnx. Done

File: config/config.go [hunk 10]
```diff
@@ -17,7 +19,7 @@ import (
 	// logconfig "github.com/allegro/akubra/log/config"
 	// "github.com/allegro/akubra/metrics"
 	// shardingconfig "github.com/allegro/akubra/sharding/config"
```
Comment (janisz) [inline config/config.go]: Commented code

File: config/config.go [hunk 11]
```diff
@@ -75,34 +72,22 @@ func Configure(configFilePath string) (conf Config, err error) {
...
 	valid, validationErrors := validator.Validate(conf)
+	println("valid so far", valid)
 	if valid && enableLogicalValidator {
 		var validListenPorts bool
-		conf.RegionsEntryLogicalValidator(&valid, &validationErrors)
 		conf.ListenPortsLogicalValidator(&validListenPorts, &validationErrors)
+		conf.RegionsEntryLogicalValidator(&valid, &validationErrors)
+
 		valid = valid && validListenPorts
+		println("logic valid", valid, validListenPorts)
```
Comment (janisz) [inline config/config.go]: why not `log.Println`?

File: config/validator.go [hunk 12]
```diff
@@ -59,54 +60,57 @@ func UniqueValuesInSliceValidator(v interface{}, param string) error {
...
-	// 	errorDetail := []error{errors.New("Listen and TechnicalEndpointListen has the same port")}
-	// 	errorsList["ListenPortsLogicalValidator"] = errorDetail
-	// } else {
-	// 	*valid = true
-	// }
-	// *validationErrors = mergeErrors(*validationErrors, errorsList)
+	errorsList := make(map[string][]error)
+	listenParts := strings.Split(c.Service.Server.Listen, ":")
+	listenTechnicalParts := strings.Split(c.Service.Server.TechnicalEndpointListen, ":")
+	*valid = true
```
Comment (janisz) [inline config/validator.go]: what if `valid == false`?

File: crdstore/config/config.go [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"github.com/allegro/akubra/metrics"
+	"github.com/allegro/akubra/types"
+)
+
+// CredentialsStore - config for credential store
+type CredentialsStore struct {
+	Endpoint types.YAMLUrl    `yaml:"Endpoint"`
+	CacheTTL metrics.Interval `yaml:"CacheTTL"`
```
Comment (janisz) [inline crdstore/config/config.go]: Add docstring

File: crdstore/config/config.go [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+// CredentialsStoreMap - map of credentialsStores configurations
+type CredentialsStoreMap map[string]CredentialsStore
```
Comment (janisz) [inline crdstore/config/config.go]: what is the key in this map?

File: crdstore/crdstore.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+	cs.cache.Store(key, newCsd)
+	cs.lock.Unlock()
+	if newCsd.AccessKey == "" {
+		return nil, newCsd.err
+	}
+	return newCsd, nil
+}
+
+func (cs *CredentialsStore) tryLock() bool {
+	return atomic.CompareAndSwapInt32((*int32)(unsafe.Pointer(&cs.lock)), 0, 1)
```
Comment (janisz) [inline crdstore/crdstore.go]: why do we need it? Add a comment

File: crdstore/crdstore.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+		csd = value.(*CredentialsStoreData)
+	}
+
+	switch {
+	case csd == nil || csd.AccessKey == "":
+		return cs.updateCache(accessKey, backend, key, csd, true)
+	case time.Now().After(csd.EOL):
+		return cs.updateCache(accessKey, backend, key, csd, false)
+	case time.Now().Add(cs.TTL / 100 * (100 - refreshTTLPercent)).After(csd.EOL):
+		go cs.updateCache(accessKey, backend, key, csd, false)
```
Comment (janisz) [inline crdstore/crdstore.go]: potentially dangerous. You do not control number of gorutines

File: crdstore/crdstore.go (hunk 15 above)
Comment (mjarco) [reply crdstore/crdstore.go]: For non blocking credentials update operations. Open question is If we need those. I doubt, so we'll clear things up before merge to master

**Pull request merged**